### DISCUSSION TITLE
Question: does meson-python support python namespaces

### DISCUSSION THREAD

**Discussion by matteoaletti** [mesonbuild/meson-python#582] [Q&A] [answered]
I am trying to migrate a project with **python namespaces** from setuptools to meson. 

I package it with setuptools with this option in the pyproject.toml

> **[tool.setuptools.packages.find]**
> namespaces = True

While trying meson as build backend, I was able to have a working wheel only without the namespace structure. When I tried with the namespace structure I could not import the package because it was not finding the namespace.

Does meson support them?

**Comment by rgommers** [comment, marked as answer]
Hi @matteoaletti. I'm not 100% sure, but I think that using namespace packages shouldn't be a problem. We don't have a specific test for it, but if you look at https://packaging.python.org/en/latest/guides/packaging-namespace-packages/, then you'll see that all there is to it (from a package build perspective) is having a top-level directory without an `__init__.py` in it. And since Meson lets you explicitly control which files to install and to where, I would not expect any interesting differences between a regular and a namespace package.

I don't quite know what `[tool.setuptools.packages.find]` does - probably some remapping of install directories? Do you know, and can you share your code? If not, could you please try to share a minimal/dummy package that is representative of your problem?

**Comment by agriyakhetarpal** [comment]
> I don't quite know what [tool.setuptools.packages.find] does - probably some remapping of install directories?

I can take a stab at answering this. With `[tool.setuptools.packages.find]`, `setuptools` shall find namespace packages by default; with PEP517/518 the behaviour for `pyproject.toml`-based projects (`setuptools.build_meta`) is that it can emit a "Package would be ignored" warning but still includes it in the list of paths that get copied and installed subsequently. This is because `include_package_data=True` is enabled by default for projects with `pyproject.toml`-based declarative metadata, but this is disabled and set to `False` for `setup.py`-based builds or those with the `setuptools.build_meta.__legacy__` build-backend (I am not aware of the reasons behind this).

`find = {namespaces = false}` should disable the lookup of namespace packages, though this is with `[tool.setuptools.packages]`, not through `[tool.setuptools.packages.find]` (both of which just refer to the same thing, I guess, because of how TOML tables are written).

I usually end up using glob-based matching to counteract this, which is more explicit, neat, and works for both usual and namespace packages – and I would imagine Meson does, or should, behave in a similar way:

```toml
[tool.setuptools.packages.find]
where = ["myproject", "myproject.*"]  # for flat layouts
```

which is why it is (almost) always better to use src layouts

**Comment by rgommers** [comment]
Thanks for the context @agriyakhetarpal.

I'll convert this to a Discussion topic, since it's a usage question rather than a bug report. @matteoaletti if your question is answered please accept the answer in the UI. And if there's more to discuss, by all means do so.

**Reply by matteoaletti** [reply]
Thank you for all the quick answers. We are trying to build a simple namespace-package and after your feedback I think we made some progress. It seems the short answer to my question is "yes: python namespaces are supported".
In case of troubles, I will upload a concrete example of how we are doing it.
Thank you again for the support!

**Reply by benoit9126** [reply]
Hi @rgommers @matteoaletti I am trying to use implicit namespaces too, and I meet a problem that I tried to capture in the small example (that, I hope, will be reproducible). 

I have developed a package (named `roseau-load-flow`) that uses an implicit namespace (called `roseau`). Now, I am trying to develop a second package using the same namespace. Let's call it `roseau-bug`. In this second package, I will use `meson-python` to compile an extension.

Here are the files of this small example: [Roseau_Bug.zip](https://github.com/user-attachments/files/17247387/Roseau_Bug.zip)


After unzipping the archive, the structure is the following (I reuse the example of the [Tutorial](https://mesonbuild.com/meson-python/tutorials/introduction.html))

```bash
> tree .
.
├── meson.build
├── pyproject.toml
├── README.md
└── roseau
    └── bug
        ├── __init__.py
        ├── meson.build
        └── our_first_module.c
```

Now I use the following commands:

```bash
 # Pin a Python version
rye pin 3.12

# Install the dependency and build an editable install of the current repository
rye sync

# Try to import an element from the editable install
python -c "from roseau.bug import our_first_module; print(our_first_module.foo())" # Works (print "bar")

# Try to import an element coming from the same namespace (but from a dependency of the current project)
python -c "import roseau.load_flow  as rlf" # Fails
```
The last command fails with:
```
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'roseau.load_flow'
```

When I look in the `.venv/`  folder created by rye, `roseau.load_flow` module seems to be well installed. As far as I understand, the file `_roseau_bug_editable_loader.py` generated by meson-python seems to have completely shadowed installed dependencies that live in the same namespace...

When I manually delete the files `roseau-bug-editable.pth` and `_roseau_bug_editable_loader.py`, I can import `roseau.load_flow` module without problem:
```bash
rm .venv/lib/python3.12/site-packages/roseau-bug-editable.pth 
rm .venv/lib/python3.12/site-packages/_roseau_bug_editable_loader.py
python -c "import roseau.load_flow  as rlf" # Works now
```

Is there something that I haven't understood in my usage of namespace with meson-python? Thanks for your help!

**Reply by dnicolodi** [reply]
Thanks @benoit9126 for reporting this issue. It seems to be a problem of the editable wheel loader. I'll check what we can do to fix it, but it may be that the system we use to implement editable wheel is inherently not compatible with namespace packages.

**Reply by benoit9126** [reply]
Thanks @dnicolodi for your help. It would be really nice to be able to develop packages using meson-python with dependencies in the same namespace.

**Reply by benoit9126** [reply]
@dnicolodi I think I have found a possible mistake in the editable loader.

In the function `find_spec` of the editable loader, there is this code:

```python
def find_spec(fullname: str, tree: Node) -> Optional[importlib.machinery.ModuleSpec]:
    namespace = False
    parts = fullname.split('.')

    # look for a package
    package = tree.get(tuple(parts))
    if isinstance(package, Node):
        for loader, suffix in LOADERS:
            src = package.get('__init__' + suffix)
            if isinstance(src, str):
                return build_module_spec(loader, fullname, src, package)
        else:
            namespace = True

    # look for a module
    for loader, suffix in LOADERS:
        src = tree.get((*parts[:-1], parts[-1] + suffix))
        if isinstance(src, str):
            return build_module_spec(loader, fullname, src, None)

    # namespace
    if namespace:
        spec = importlib.machinery.ModuleSpec(fullname, None, is_package=True)
        assert isinstance(spec.submodule_search_locations, list)  # make mypy happy
        spec.submodule_search_locations.append(os.path.join(__file__, fullname))
        return spec

    return None
```

If you look carefully at the line `spec.submodule_search_locations.append(os.path.join(__file__, fullname))` in the `if namespace:` section, you will realize that the path concatenation give a weird result (in my opinion).

`__file__` contains the name of the current Python file. In my example `_roseau_bug_editable_load_file.py` and `fullname` contains the name of the package to import (which is a namespace in this section). In my example, it is `roseau`. So the concatenation of both gives something like:
`'/<something>/.venv/lib/python3.12/site-packages/_roseau_bug_editable_loader.py/roseau'` which is not a valid path.

I would replace this line with
```python
spec.submodule_search_locations.append(os.path.join(os.path.dirname(__file__), fullname))
```
or even
```python
spec.submodule_search_locations.append(os.path.join(os.path.dirname(__file__), fullname.replace(".",os.path.sep)))
```
to ensure valid path.

In my example, the first patch works.

**Reply by dnicolodi** [reply]
Thanks @benoit9126 for looking into this. I didn't really find enough time to analyze the details. I have vague memories that the weird path is needed to support some other test case, but I may be misremembering. I should have places a comment there... Can you verify that the test suite still passes with your change?

**Reply by benoit9126** [reply]
@dnicolodi I will try to prepare a PR with this change. Problems may appear with the CI.
